### PR TITLE
Add NaiveDirichlet and NaiveBeta distributions for testing

### PR DESCRIPTION
**Pull request by fritzo**
This adds two new distributions with noisy reparameterized gradients for testing. This also refactors our distribution tests a bit to allow testing of stateful distributions which cannot implement a functional form via `RandomPrimitive`. Distribution tests now include:
- `NaiveDirichlet`
- `NaiveBeta`
- `ShapeAugmentedGamma` which tests `Rejector`
- `RejectionExponential` which tests `Rejector`

This figure shows the difference between `Beta` "canonical" and `NaiveBeta` "reptrick":
beta-motivation

### CODE REVIEW

**Comment by fritzo** [comment]
cc @martinjankowiak

**Comment by fritzo** [comment]
Hmm, still getting this to work, please don't merge yet.

**Comment by fritzo** [comment]
Ok now it's working, ready to merge.

**Comment by fritzo** [comment]
@martinjankowiak would you mind merging this?

File: tests/distributions/test_distributions.py [hunk 1]
```diff
@@ -27,34 +27,36 @@ def _log_prob_shape(dist, x_size=torch.Size(), **dist_params):
...
         dist_params = dist.get_dist_params(idx)
+        d = dist.pyro_dist_class(**dist_params)
         x = dist.get_test_data(idx)
         with xfail_if_not_implemented():
             # Get log_prob shape after broadcasting.
-            expected_shape = _log_prob_shape(d, x.size(), **dist_params)
-            log_p_func = d.log_prob(x, **dist_params)
-            log_p_obj = dist.pyro_dist_obj(**dist_params).log_prob(x)
+            expected_shape = _log_prob_shape(d, x.size())
+            log_p_func = d.log_prob(x)
```
Comment (neerajprad) [inline tests/distributions/test_distributions.py]: This is the same as `log_p_obj` now. While checking for consistency between RandomPrimitive and Distribution instances isn't that important now, it ensures that we are calling the RandomPrimitive code path for all major methods without throwing exceptions. Can we just do 
```
d=dist.pyro_dist_class(**dist_params) if dist.pyro_dist is None else dist.pyro_dist
```
, or remove `log_p_func` and redundant checks?

Comment (fritzo) [reply tests/distributions/test_distributions.py]: Done.

**Pull request merged**